### PR TITLE
[SPIR-V] Add pre-commit CI workflow

### PR DESCRIPTION
**Pull request by sudonatalie**
Add a pre-commit CI workflow for the experimental SPIR-V backend. This action should only run when SPIR-V target or test files are modified. The `codegen-spirv` tests don't run as part of `check-all` because the SPIR-V backend is still experimental.

Depends on #73371 (for a green tree)

### CODE REVIEW

**Review by tstellar — COMMENTED** [review]
We depend a lot on ccache to get faster builds, do you know if adding `DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=spirv` makes any changes to header files that may invalidate the ccache data?

Would it make sense to start by having this job manually triggered either by a comment or a label?

How important is it to run this job on Windows and Mac OS?  We are resource constrained on those OSes.

File: .github/workflows/spirv-tests.yml
Comment (sudonatalie) [reply .github/workflows/spirv-tests.yml]: @tstellar Thanks for taking a look!

> We depend a lot on ccache to get faster builds, do you know if adding `DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=spirv` makes any changes to header files that may invalidate the ccache data?

Do you have any suggestions for how I'd go about checking this? I've never used ccache before, but I just ran a couple tests with it using separate clean build directories and `-DCMAKE_CXX_COMPILER=/usr/lib/ccache/clang++ -DCMAKE_C_COMPILER=/usr/lib/ccache/clang` but the builds take ~the same time with the same CMake args regardless of whether the cache is cleaned or populated.

> Would it make sense to start by having this job manually triggered either by a comment or a label?

In my opinion, I think the benefit of these checks would be that they'd run automatically, so people touching SPIR-V backend files get that feedback on their PR even if they forget/don't know that this bot exists. However, I understand we're using shared resources here, so if a manual trigger is a necessary compromise to start with then I'd certainly rather that than nothing.

> How important is it to run this job on Windows and Mac OS? We are resource constrained on those OSes.

Much less important since the behavior shouldn't diverge on different platforms, so Linux only is fine. Still working on figuring out how to configure that with GitHub actions though, my first several attempts to override `runs-on` from the calling workflow were unsuccessful so I'm going to try to find a better way to test it than repeatedly force-pushing here.

File: .github/workflows/llvm-project-tests.yml
Comment (sudonatalie) [reply .github/workflows/llvm-project-tests.yml]: Done

**Comment by sudonatalie** [comment]
@tstellar Thanks for taking a look!

> We depend a lot on ccache to get faster builds, do you know if adding `DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD=spirv` makes any changes to header files that may invalidate the ccache data?

Do you have any suggestions for how I'd go about checking this? I've never used ccache before, but I just ran a couple tests with it using separate clean build directories and `-DCMAKE_CXX_COMPILER=/usr/lib/ccache/clang++ -DCMAKE_C_COMPILER=/usr/lib/ccache/clang` but the builds take ~the same time with the same CMake args regardless of whether the cache is cleaned or populated.

> Would it make sense to start by having this job manually triggered either by a comment or a label?

In my opinion, I think the benefit of these checks would be that they'd run automatically, so people touching SPIR-V backend files get that feedback on their PR even if they forget/don't know that this bot exists. However, I understand we're using shared resources here, so if a manual trigger is a necessary compromise to start with then I'd certainly rather that than nothing.

> How important is it to run this job on Windows and Mac OS? We are resource constrained on those OSes.

Much less important since the behavior shouldn't diverge on different platforms, so Linux only is fine. Still working on figuring out how to configure that with GitHub actions though, my first several attempts to override `runs-on` from the calling workflow were unsuccessful so I'm going to try to find a better way to test it than repeatedly force-pushing here.

File: .github/workflows/spirv-tests.yml
Comment (sudonatalie) [reply .github/workflows/spirv-tests.yml]: Done (assuming `-DLLVM_TARGETS_TO_BUILD=""` was the right way to do that)

File: .github/workflows/llvm-project-tests.yml
Comment (sudonatalie) [inline .github/workflows/llvm-project-tests.yml:56]: This does feel a bit hacky but it's the best solution I found for parameterizing `matrix.os`.

**Comment by tstellar** [comment]
> Do you have any suggestions for how I'd go about checking this? I've never used ccache before, but I just ran a couple tests with it using separate clean build directories and `-DCMAKE_CXX_COMPILER=/usr/lib/ccache/clang++ -DCMAKE_C_COMPILER=/usr/lib/ccache/clang` but the builds take ~the same time with the same CMake args regardless of whether the cache is cleaned or populated.

I think the way to test it would be:
```
# Configure with default targets:
cmake -G Ninja -S llvm -B build -DCMAKE_BUILD_TYPE=Release -DCMAKE_CXX_COMPILER=/usr/lib/ccache/clang++ -DCMAKE_C_COMPILER=/usr/lib/ccache/clang

# Build llvm:
ninja -C build

# Save ccache stats
ccache -s > phase1.stats

# Rebuild llvm:
ninja -C build

# Save ccache stats
ccache -s > phase2.stats

# Edit CMakeCache.txt and enable SPIRV
vim build/CMakeCache.txt
-> LLVM_EXPERIMENTAL_TARGETS_TO_BUILD:STRING=SPIRV

# Rebuild llvm:
ninja -C build

# Save ccache stats:
ccache -s > phase3.stats
```
If it worked, when you compare the phase2.stats and phase3.stats output it should should mostly an increase in the Hits category.  I did this test locally and it appears that adding -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD:STRING=SPIRV does not invalidate the cache, which is good, but the SPIRV backend failed to build, so you might want to double check my work here.

File: .github/workflows/llvm-project-tests.yml
Comment (tstellar) [reply .github/workflows/llvm-project-tests.yml:56]: This is the only way to do this as far as I know.

File: .github/workflows/spirv-tests.yml
Comment (tstellar) [inline .github/workflows/spirv-tests.yml:27]: Did you want to enable clang here still?  I think the first version had clang enabled.

Comment (sudonatalie) [reply .github/workflows/spirv-tests.yml:27]: I thought it was necessary to supply a project so I went with `clang` but empty is fine and makes the build a bit faster.

**Comment by sudonatalie** [comment]
> If it worked, when you compare the phase2.stats and phase3.stats output it should should mostly an increase in the Hits category. I did this test locally and it appears that adding -DLLVM_EXPERIMENTAL_TARGETS_TO_BUILD:STRING=SPIRV does not invalidate the cache, which is good, but the SPIRV backend failed to build, so you might want to double check my work here.

Looks the same for me and the build succeeded.

**Comment by sudonatalie** [comment]
> Looks the same for me and the build succeeded.

Actually I just pulled upstream and the SPIR-V backend is broken at tip-of-tree at the moment. Fix in #74660.

File: .github/workflows/spirv-tests.yml
Comment (tstellar) [reply .github/workflows/spirv-tests.yml:29]: OK, thanks for testing this out.  I will look into it.

**Review by tstellar — APPROVED** [review]
LGTM with the caveat that if this ends up being too resource intensive, we may need to switch it to be manually triggered or disable it until we can get more resources.

**Comment by tstellar** [comment]
@michalpaszkowski I have not seen this before.

**Comment by boomanaiden154** [comment]
We recently switched the container to use a non-root user by default. I'm guessing that is what's causing the issue as the default user will not have permission to install applications.

**Comment by VyacheslavLevytskyy** [comment]
This now always leads to fails in Github Actions for linux runs, and this doesn't depend on the sub-project. I can see the same fail here, for example: https://github.com/llvm/llvm-project/actions/runs/12631817480/job/35194267947

I guess it's a blocker now that causes failures in every sub-project that relies on `.github/workflows/llvm-project-tests.yml`. @boomanaiden154 @tstellar May I ask your advice, what you consider the best way forward to fix/mitigate this?

**Comment by boomanaiden154** [comment]
It's probably better to just add the `gha` to the sudoers file and give it passwordless sudo access and then modify the other actions to install things with sudo. I can work on a patch and hopefully get it up by EOD.

**Comment by VyacheslavLevytskyy** [comment]
> give the container user write access to /usr/local/

@boomanaiden154 Can we give the container user write access to /usr/local/ to proceed with the simplest solution?

**Comment by boomanaiden154** [comment]
> May I ask you what do you think about this idea to give the container user write access to /usr/local/ ?

#122221 already fixed the issue.

**Comment by VyacheslavLevytskyy** [comment]
> > May I ask you what do you think about this idea to give the container user write access to /usr/local/ ?
> 
> #122221 already fixed the issue.

Thank you!

**Pull request merged**
